File: tests/compression_tests/text_compression/text_compression.go
package main

import (
	"bytes"
	"compress/gzip"
	"compress/zlib"
	"encoding/json"
	"fmt"
	"io"
	"math/rand"
	"os"
	"strings"
	"time"
)

type CompressionResult struct {
	Success         bool    `json:"success"`
	CompressedSize  *int    `json:"compressed_size,omitempty"`
	CompressionTime float64 `json:"compression_time"`
	Error           *string `json:"error,omitempty"`
}

type DecompressionResult struct {
	Success           bool    `json:"success"`
	DecompressedSize  *int    `json:"decompressed_size,omitempty"`
	DecompressionTime float64 `json:"decompression_time"`
	Error             *string `json:"error,omitempty"`
}

type IterationResult struct {
	Iteration     int                  `json:"iteration"`
	OriginalSize  int                  `json:"original_size"`
	Compression   CompressionResult    `json:"compression"`
	Decompression *DecompressionResult `json:"decompression,omitempty"`
}

type TestCase struct {
	InputSize            int               `json:"input_size"`
	TextType             string            `json:"text_type"`
	Algorithm            string            `json:"algorithm"`
	Iterations           []IterationResult `json:"iterations"`
	AvgCompressionRatio  float64           `json:"avg_compression_ratio"`
	AvgCompressionTime   float64           `json:"avg_compression_time"`
	AvgDecompressionTime float64           `json:"avg_decompression_time"`
}

type AlgorithmPerformance struct {
	AvgCompressionRatio float64 `json:"avg_compression_ratio"`
	MaxCompressionRatio float64 `json:"max_compression_ratio"`
	MinCompressionRatio float64 `json:"min_compression_ratio"`
}

type Summary struct {
	TotalTests             int                             `json:"total_tests"`
	SuccessfulCompressions int                             `json:"successful_compressions"`
	FailedCompressions     int                             `json:"failed_compressions"`
	BestCompressionRatios  map[string]float64              `json:"best_compression_ratios"`
	AlgorithmPerformance   map[string]AlgorithmPerformance `json:"algorithm_performance"`
}

type BenchmarkResults struct {
	StartTime          float64    `json:"start_time"`
	TestCases          []TestCase `json:"test_cases"`
	Summary            Summary    `json:"summary"`
	EndTime            *float64   `json:"end_time,omitempty"`
	TotalExecutionTime *float64   `json:"total_execution_time,omitempty"`
}

type Config struct {
	Parameters Parameters `json:"parameters"`
}

type Parameters struct {
	InputSizes            []int    `json:"input_sizes"`
	TextTypes             []string `json:"text_types"`
	CompressionAlgorithms []string `json:"compression_algorithms"`
	Iterations            int      `json:"iterations"`
}

func generateTextData(size int, textType string) (string, error) {
	rand.Seed(time.Now().UnixNano())

	switch textType {
	case "ascii":
		chars := "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789 \n"
		result := make([]byte, size)
		for i := 0; i < size; i++ {
			result[i] = chars[rand.Intn(len(chars))]
		}
		return string(result), nil

	case "unicode":
		chars := "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789àáâãäåæçèéêëìíîïñòóôõöøùúûüý你好世界🌟🚀📊 \n"
		runes := []rune(chars)
		result := make([]rune, 0, size)
		for len(string(result)) < size {
			result = append(result, runes[rand.Intn(len(runes))])
		}
		resultStr := string(result)
		if len(resultStr) > size {
			// Truncate to exact size, being careful with UTF-8
			return resultStr[:size], nil
		}
		return resultStr, nil

	case "code":
		keywords := []string{"package", "func", "var", "if", "else", "for", "range", "return", "struct", "interface"}
		operators := []string{"=", "+", "-", "*", "/", "(", ")", "{", "}", "[", "]", ";", ":"}
		var text strings.Builder

		for text.Len() < size {
			if rand.Float64() < 0.3 {
				text.WriteString(keywords[rand.Intn(len(keywords))])
			} else {
				wordLen := rand.Intn(8) + 3
				for i := 0; i < wordLen; i++ {
					text.WriteByte(byte('a' + rand.Intn(26)))
				}
			}

			if rand.Float64() < 0.2 {
				text.WriteString(operators[rand.Intn(len(operators))])
			}

			if rand.Float64() < 0.1 {
				text.WriteString("\n")
			} else {
				text.WriteString(" ")
			}
		}

		result := text.String()
		if len(result) > size {
			return result[:size], nil
		}
		return result, nil

	case "natural_language":
		words := []string{"the", "quick", "brown", "fox", "jumps", "over", "lazy", "dog", "and", "runs", "through",
			"forest", "meadow", "river", "mountain", "valley", "beautiful", "magnificent", "wonderful"}
		var text strings.Builder

		for text.Len() < size {
			text.WriteString(words[rand.Intn(len(words))])

			if rand.Float64() < 0.1 {
				text.WriteString(". ")
			} else if rand.Float64() < 0.05 {
				text.WriteString(", ")
			} else {
				text.WriteString(" ")
			}

			if rand.Float64() < 0.05 {
				text.WriteString("\n")
			}
		}

		result := text.String()
		if len(result) > size {
			return result[:size], nil
		}
		return result, nil

	default:
		return "", fmt.Errorf("unknown text type: %s", textType)
	}
}

func compressWithGzip(data []byte) CompressionResult {
	start := time.Now()

	var buf bytes.Buffer
	writer := gzip.NewWriter(&buf)

	_, err := writer.Write(data)
	if err != nil {
		errStr := err.Error()
		return CompressionResult{
			Success:         false,
			CompressionTime: float64(time.Since(start).Nanoseconds()) / 1e6,
			Error:           &errStr,
		}
	}

	err = writer.Close()
	if err != nil {
		errStr := err.Error()
		return CompressionResult{
			Success:         false,
			CompressionTime: float64(time.Since(start).Nanoseconds()) / 1e6,
			Error:           &errStr,
		}
	}

	compressed := buf.Bytes()
	compressedSize := len(compressed)

	return CompressionResult{
		Success:         true,
		CompressedSize:  &compressedSize,
		CompressionTime: float64(time.Since(start).Nanoseconds()) / 1e6,
	}
}

func compressWithZlib(data []byte) CompressionResult {
	start := time.Now()

	var buf bytes.Buffer
	writer := zlib.NewWriter(&buf)

	_, err := writer.Write(data)
	if err != nil {
		errStr := err.Error()
		return CompressionResult{
			Success:         false,
			CompressionTime: float64(time.Since(start).Nanoseconds()) / 1e6,
			Error:           &errStr,
		}
	}

	err = writer.Close()
	if err != nil {
		errStr := err.Error()
		return CompressionResult{
			Success:         false,
			CompressionTime: float64(time.Since(start).Nanoseconds()) / 1e6,
			Error:           &errStr,
		}
	}

	compressed := buf.Bytes()
	compressedSize := len(compressed)

	return CompressionResult{
		Success:         true,
		CompressedSize:  &compressedSize,
		CompressionTime: float64(time.Since(start).Nanoseconds()) / 1e6,
	}
}

func decompressGzip(data []byte) DecompressionResult {
	start := time.Now()

	reader, err := gzip.NewReader(bytes.NewReader(data))
	if err != nil {
		errStr := err.Error()
		return DecompressionResult{
			Success:           false,
			DecompressionTime: float64(time.Since(start).Nanoseconds()) / 1e6,
			Error:             &errStr,
		}
	}
	defer reader.Close()

	decompressed, err := io.ReadAll(reader)
	if err != nil {
		errStr := err.Error()
		return DecompressionResult{
			Success:           false,
			DecompressionTime: float64(time.Since(start).Nanoseconds()) / 1e6,
			Error:             &errStr,
		}
	}

	decompressedSize := len(decompressed)

	return DecompressionResult{
		Success:           true,
		DecompressedSize:  &decompressedSize,
		DecompressionTime: float64(time.Since(start).Nanoseconds()) / 1e6,
	}
}

func decompressZlib(data []byte) DecompressionResult {
	start := time.Now()

	reader, err := zlib.NewReader(bytes.NewReader(data))
	if err != nil {
		errStr := err.Error()
		return DecompressionResult{
			Success:           false,
			DecompressionTime: float64(time.Since(start).Nanoseconds()) / 1e6,
			Error:             &errStr,
		}
	}
	defer reader.Close()

	decompressed, err := io.ReadAll(reader)
	if err != nil {
		errStr := err.Error()
		return DecompressionResult{
			Success:           false,
			DecompressionTime: float64(time.Since(start).Nanoseconds()) / 1e6,
			Error:             &errStr,
		}
	}

	decompressedSize := len(decompressed)

	return DecompressionResult{
		Success:           true,
		DecompressedSize:  &decompressedSize,
		DecompressionTime: float64(time.Since(start).Nanoseconds()) / 1e6,
	}
}

func runTextCompressionBenchmark(config Parameters) (BenchmarkResults, error) {
	inputSizes := config.InputSizes
	if len(inputSizes) == 0 {
		inputSizes = []int{1024}
	}

	textTypes := config.TextTypes
	if len(textTypes) == 0 {
		textTypes = []string{"ascii"}
	}

	algorithms := config.CompressionAlgorithms
	if len(algorithms) == 0 {
		algorithms = []string{"gzip"}
	}

	iterations := config.Iterations
	if iterations == 0 {
		iterations = 3
	}

	results := BenchmarkResults{
		StartTime: float64(time.Now().Unix()),
		TestCases: []TestCase{},
		Summary: Summary{
			BestCompressionRatios: make(map[string]float64),
			AlgorithmPerformance:  make(map[string]AlgorithmPerformance),
		},
	}

	algorithmStats := make(map[string][]float64)

	for _, size := range inputSizes {
		for _, textType := range textTypes {
			for _, algorithm := range algorithms {
				fmt.Fprintf(os.Stderr, "Testing %s text, size: %d, algorithm: %s...\n", textType, size, algorithm)

				testCase := TestCase{
					InputSize:  size,
					TextType:   textType,
					Algorithm:  algorithm,
					Iterations: []IterationResult{},
				}

				var compressionRatios []float64
				var compressionTimes []float64
				var decompressionTimes []float64

				for i := 0; i < iterations; i++ {
					fmt.Fprintf(os.Stderr, "  Iteration %d/%d...\n", i+1, iterations)

					textData, err := generateTextData(size, textType)
					if err != nil {
						return results, err
					}

					dataBytes := []byte(textData)
					originalSize := len(dataBytes)

					var compressResult CompressionResult

					switch algorithm {
					case "gzip":
						compressResult = compressWithGzip(dataBytes)
					case "zlib":
						compressResult = compressWithZlib(dataBytes)
					default:
						fmt.Fprintf(os.Stderr, "Warning: Algorithm %s not implemented, skipping\n", algorithm)
						continue
					}

					iterationResult := IterationResult{
						Iteration:    i + 1,
						OriginalSize: originalSize,
						Compression:  compressResult,
					}

					results.Summary.TotalTests++

					if compressResult.Success && compressResult.CompressedSize != nil {
						results.Summary.SuccessfulCompressions++

						compressedSize := *compressResult.CompressedSize
						var compressionRatio float64
						if compressedSize > 0 {
							compressionRatio = float64(originalSize) / float64(compressedSize)
						}

						compressionRatios = append(compressionRatios, compressionRatio)
						compressionTimes = append(compressionTimes, compressResult.CompressionTime)

						algorithmStats[algorithm] = append(algorithmStats[algorithm], compressionRatio)
					} else {
						results.Summary.FailedCompressions++
					}

					testCase.Iterations = append(testCase.Iterations, iterationResult)
				}

				if len(compressionRatios) > 0 {
					sum := 0.0
					for _, ratio := range compressionRatios {
						sum += ratio
					}
					testCase.AvgCompressionRatio = sum / float64(len(compressionRatios))

					sum = 0.0
					for _, time := range compressionTimes {
						sum += time
					}
					testCase.AvgCompressionTime = sum / float64(len(compressionTimes))

					if len(decompressionTimes) > 0 {
						sum = 0.0
						for _, time := range decompressionTimes {
							sum += time
						}
						testCase.AvgDecompressionTime = sum / float64(len(decompressionTimes))
					}
				}

				results.TestCases = append(results.TestCases, testCase)
			}
		}
	}

	// Calculate summary statistics
	for algorithm, ratios := range algorithmStats {
		if len(ratios) > 0 {
			sum := 0.0
			min := ratios[0]
			max := ratios[0]

			for _, ratio := range ratios {
				sum += ratio
				if ratio < min {
					min = ratio
				}
				if ratio > max {
					max = ratio
				}
			}

			results.Summary.AlgorithmPerformance[algorithm] = AlgorithmPerformance{
				AvgCompressionRatio: sum / float64(len(ratios)),
				MaxCompressionRatio: max,
				MinCompressionRatio: min,
			}
		}
	}

	endTime := float64(time.Now().Unix())
	results.EndTime = &endTime
	totalTime := endTime - results.StartTime
	results.TotalExecutionTime = &totalTime

	return results, nil
}

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintf(os.Stderr, "Usage: %s <config_file>\n", os.Args[0])
		os.Exit(1)
	}

	configFile := os.Args[1]

	configData, err := os.ReadFile(configFile)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: Config file '%s' not found\n", configFile)
		os.Exit(1)
	}

	var config Config
	err = json.Unmarshal(configData, &config)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: Invalid JSON in config file: %v\n", err)
		os.Exit(1)
	}

	results, err := runTextCompressionBenchmark(config.Parameters)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}

	output, err := json.MarshalIndent(results, "", "  ")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}

	fmt.Println(string(output))
}

File: tests/compression_tests/text_compression/text_compression.rs
use std::env;
use std::fs;
use std::time::Instant;
use std::collections::HashMap;
use serde::{Deserialize, Serialize};
use serde_json;
use rand::Rng;
use rand::seq::SliceRandom;
use flate2::write::GzEncoder;
use flate2::write::ZlibEncoder;
use flate2::Compression;
use std::io::Write;

#[derive(Debug, Serialize)]
struct CompressionResult {
    success: bool,
    compressed_size: Option<usize>,
    compression_time: f64,
    error: Option<String>,
}

#[derive(Debug, Serialize)]
struct DecompressionResult {
    success: bool,
    decompressed_size: Option<usize>,
    decompression_time: f64,
    error: Option<String>,
}

#[derive(Debug, Serialize)]
struct IterationResult {
    iteration: u32,
    original_size: usize,
    compression: CompressionResult,
    decompression: Option<DecompressionResult>,
}

#[derive(Debug, Serialize)]
struct TestCase {
    input_size: usize,
    text_type: String,
    algorithm: String,
    iterations: Vec<IterationResult>,
    avg_compression_ratio: f64,
    avg_compression_time: f64,
    avg_decompression_time: f64,
}

#[derive(Debug, Serialize)]
struct AlgorithmPerformance {
    avg_compression_ratio: f64,
    max_compression_ratio: f64,
    min_compression_ratio: f64,
}

#[derive(Debug, Serialize)]
struct Summary {
    total_tests: u32,
    successful_compressions: u32,
    failed_compressions: u32,
    best_compression_ratios: HashMap<String, f64>,
    algorithm_performance: HashMap<String, AlgorithmPerformance>,
}

#[derive(Debug, Serialize)]
struct BenchmarkResults {
    start_time: f64,
    test_cases: Vec<TestCase>,
    summary: Summary,
    end_time: Option<f64>,
    total_execution_time: Option<f64>,
}

#[derive(Debug, Deserialize)]
struct Config {
    parameters: Parameters,
}

#[derive(Debug, Deserialize)]
struct Parameters {
    input_sizes: Option<Vec<usize>>,
    text_types: Option<Vec<String>>,
    compression_algorithms: Option<Vec<String>>,
    iterations: Option<u32>,
}

fn generate_text_data(size: usize, text_type: &str) -> Result<String, String> {
    let mut rng = rand::thread_rng();
    
    match text_type {
        "ascii" => {
            let chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789 \n";
            let result: String = (0..size)
                .map(|_| chars.chars().nth(rng.gen_range(0..chars.len())).unwrap())
                .collect();
            Ok(result)
        }
        "unicode" => {
            let chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789àáâãäåæçèéêëìíîïñòóôõöøùúûüý你好世界🌟🚀📊 \n";
            let result: String = (0..size)
                .map(|_| chars.chars().nth(rng.gen_range(0..chars.chars().count())).unwrap())
                .collect();
            Ok(result)
        }
        "code" => {
            let keywords = vec!["fn", "let", "mut", "if", "else", "for", "while", "return", "match", "impl"];
            let operators = vec!["=", "+", "-", "*", "/", "(", ")", "{", "}", "[", "]", ";", ":"];
            let mut text = String::new();
            
            while text.len() < size {
                if rng.gen::<f64>() < 0.3 {
                    text.push_str(keywords.choose(&mut rng).unwrap());
                } else {
                    let word_len = rng.gen_range(3..10);
                    for _ in 0..word_len {
                        text.push((b'a' + rng.gen_range(0..26)) as char);
                    }
                }
                
                if rng.gen::<f64>() < 0.2 {
                    text.push_str(operators.choose(&mut rng).unwrap());
                }
                
                if rng.gen::<f64>() < 0.1 {
                    text.push('\n');
                } else {
                    text.push(' ');
                }
            }
            
            Ok(text.chars().take(size).collect())
        }
        "natural_language" => {
            let words = vec!["the", "quick", "brown", "fox", "jumps", "over", "lazy", "dog", "and", "runs", "through",
                           "forest", "meadow", "river", "mountain", "valley", "beautiful", "magnificent", "wonderful"];
            let mut text = String::new();
            
            while text.len() < size {
                text.push_str(words.choose(&mut rng).unwrap());
                
                if rng.gen::<f64>() < 0.1 {
                    text.push_str(". ");
                } else if rng.gen::<f64>() < 0.05 {
                    text.push_str(", ");
                } else {
                    text.push(' ');
                }
                
                if rng.gen::<f64>() < 0.05 {
                    text.push('\n');
                }
            }
            
            Ok(text.chars().take(size).collect())
        }
        _ => Err(format!("Unknown text type: {}", text_type))
    }
}

fn compress_with_gzip(data: &[u8]) -> CompressionResult {
    let start = Instant::now();
    
    match try_compress_gzip(data) {
        Ok(compressed) => {
            let compression_time = start.elapsed().as_secs_f64() * 1000.0;
            CompressionResult {
                success: true,
                compressed_size: Some(compressed.len()),
                compression_time,
                error: None,
            }
        }
        Err(e) => {
            let compression_time = start.elapsed().as_secs_f64() * 1000.0;
            CompressionResult {
                success: false,
                compressed_size: None,
                compression_time,
                error: Some(e.to_string()),
            }
        }
    }
}

fn try_compress_gzip(data: &[u8]) -> Result<Vec<u8>, Box<dyn std::error::Error>> {
    let mut encoder = GzEncoder::new(Vec::new(), Compression::default());
    encoder.write_all(data)?;
    Ok(encoder.finish()?)
}

fn compress_with_zlib(data: &[u8]) -> CompressionResult {
    let start = Instant::now();
    
    match try_compress_zlib(data) {
        Ok(compressed) => {
            let compression_time = start.elapsed().as_secs_f64() * 1000.0;
            CompressionResult {
                success: true,
                compressed_size: Some(compressed.len()),
                compression_time,
                error: None,
            }
        }
        Err(e) => {
            let compression_time = start.elapsed().as_secs_f64() * 1000.0;
            CompressionResult {
                success: false,
                compressed_size: None,
                compression_time,
                error: Some(e.to_string()),
            }
        }
    }
}

fn try_compress_zlib(data: &[u8]) -> Result<Vec<u8>, Box<dyn std::error::Error>> {
    let mut encoder = ZlibEncoder::new(Vec::new(), Compression::default());
    encoder.write_all(data)?;
    Ok(encoder.finish()?)
}

fn run_text_compression_benchmark(config: &Parameters) -> Result<BenchmarkResults, Box<dyn std::error::Error>> {
    let default_sizes = vec![1024];
    let default_types = vec!["ascii".to_string()];
    let default_algorithms = vec!["gzip".to_string()];
    
    let input_sizes = config.input_sizes.as_ref().unwrap_or(&default_sizes);
    let text_types = config.text_types.as_ref().unwrap_or(&default_types);
    let algorithms = config.compression_algorithms.as_ref().unwrap_or(&default_algorithms);
    let iterations = config.iterations.unwrap_or(3);
    
    let mut results = BenchmarkResults {
        start_time: std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)?
            .as_secs_f64(),
        test_cases: Vec::new(),
        summary: Summary {
            total_tests: 0,
            successful_compressions: 0,
            failed_compressions: 0,
            best_compression_ratios: HashMap::new(),
            algorithm_performance: HashMap::new(),
        },
        end_time: None,
        total_execution_time: None,
    };
    
    let mut algorithm_stats: HashMap<String, Vec<f64>> = HashMap::new();
    
    for &size in input_sizes {
        for text_type in text_types {
            for algorithm in algorithms {
                eprintln!("Testing {} text, size: {}, algorithm: {}...", text_type, size, algorithm);
                
                let mut test_case = TestCase {
                    input_size: size,
                    text_type: text_type.clone(),
                    algorithm: algorithm.clone(),
                    iterations: Vec::new(),
                    avg_compression_ratio: 0.0,
                    avg_compression_time: 0.0,
                    avg_decompression_time: 0.0,
                };
                
                let mut compression_ratios = Vec::new();
                let mut compression_times = Vec::new();
                
                for i in 0..iterations {
                    eprintln!("  Iteration {}/{}...", i + 1, iterations);
                    
                    let text_data = generate_text_data(size, text_type)?;
                    let data_bytes = text_data.as_bytes();
                    let original_size = data_bytes.len();
                    
                    let compress_result = match algorithm.as_str() {
                        "gzip" => compress_with_gzip(data_bytes),
                        "zlib" => compress_with_zlib(data_bytes),
                        _ => {
                            eprintln!("Warning: Algorithm {} not implemented, skipping", algorithm);
                            continue;
                        }
                    };
                    
                    let iteration_result = IterationResult {
                        iteration: i + 1,
                        original_size,
                        compression: compress_result,
                        decompression: None,
                    };
                    
                    results.summary.total_tests += 1;
                    
                    if iteration_result.compression.success {
                        results.summary.successful_compressions += 1;
                        
                        if let Some(compressed_size) = iteration_result.compression.compressed_size {
                            let compression_ratio = if compressed_size > 0 {
                                original_size as f64 / compressed_size as f64
                            } else {
                                0.0
                            };
                            
                            compression_ratios.push(compression_ratio);
                            compression_times.push(iteration_result.compression.compression_time);
                            
                            algorithm_stats.entry(algorithm.clone()).or_insert_with(Vec::new).push(compression_ratio);
                        }
                    } else {
                        results.summary.failed_compressions += 1;
                    }
                    
                    test_case.iterations.push(iteration_result);
                }
                
                if !compression_ratios.is_empty() {
                    test_case.avg_compression_ratio = compression_ratios.iter().sum::<f64>() / compression_ratios.len() as f64;
                    test_case.avg_compression_time = compression_times.iter().sum::<f64>() / compression_times.len() as f64;
                }
                
                results.test_cases.push(test_case);
            }
        }
    }
    
    // Calculate summary statistics
    for (algorithm, ratios) in algorithm_stats {
        if !ratios.is_empty() {
            let avg = ratios.iter().sum::<f64>() / ratios.len() as f64;
            let max = ratios.iter().fold(f64::NEG_INFINITY, |a, &b| a.max(b));
            let min = ratios.iter().fold(f64::INFINITY, |a, &b| a.min(b));
            
            results.summary.algorithm_performance.insert(algorithm, AlgorithmPerformance {
                avg_compression_ratio: avg,
                max_compression_ratio: max,
                min_compression_ratio: min,
            });
        }
    }
    
    results.end_time = Some(
        std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)?
            .as_secs_f64()
    );
    
    if let Some(end_time) = results.end_time {
        results.total_execution_time = Some(end_time - results.start_time);
    }
    
    Ok(results)
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: {} <config_file>", args[0]);
        std::process::exit(1);
    }
    
    let config_file = &args[1];
    
    let config_data = fs::read_to_string(config_file)
        .map_err(|_| format!("Error: Config file '{}' not found", config_file))?;
    
    let config: Config = serde_json::from_str(&config_data)
        .map_err(|e| format!("Error: Invalid JSON in config file: {}", e))?;
    
    let results = run_text_compression_benchmark(&config.parameters)?;
    
    println!("{}", serde_json::to_string_pretty(&results)?);
    
    Ok(())
}